### PR TITLE
JAMES-3029 add test to check that mail is well received by all recipients

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 1]
```diff
@@ -128,4 +170,9 @@ default void sendUniqueMessage(SMTPMessageSender sender, String message) throws
         String uniqueMessage = message.replace("banana", "UUID " + UUID.randomUUID().toString());
         sender.sendMessageWithHeaders("bob@apache.org", JAMES_USER, uniqueMessage);
     }
+
+    default void sendUniqueMessageToTwoUser(SMTPMessageSender sender, String message) throws IOException {
```
Comment (rouazana) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java]: s/User/Users

File: server/testing/src/main/java/org/apache/james/utils/SMTPMessageSender.java [hunk 2]
```diff
@@ -113,6 +114,14 @@ public SMTPMessageSender sendMessageWithHeaders(String from, String recipient, S
         return this;
     }
 
+    public SMTPMessageSender sendMessageWithHeaders(String from, List<String> recipients, String message) throws IOException {
+        doHelo();
+        doSetSender(from);
+        recipients.forEach(Throwing.consumer(this::doAddRcpt).sneakyThrow());
```
Comment (rouazana) [inline server/testing/src/main/java/org/apache/james/utils/SMTPMessageSender.java]: should sendMessageWithHeaders(String, String, String) be rewritten to use sendMessageWithHeaders(String, List<String>, String)?
bonus question: what's the difference between doRCPT and doAddRcpt?

Comment (remk) [reply server/testing/src/main/java/org/apache/james/utils/SMTPMessageSender.java]: `doAddRcpt` put the recipient into brackets, `doRCPT` not.

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 3]
```diff
@@ -89,6 +92,45 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
...
+            Mono.fromRunnable(
+                Throwing.runnable(() -> {
+                    sender.connect(JAMES_SERVER_HOST, smtpPort);
+                    sendUniqueMessageToTwoUsers(sender, message);
+                }))
+                .subscribeOn(Schedulers.elastic())
+                .block();
+        }
+
+        CALMLY_AWAIT.until(() -> server.getProbe(SpoolerProbe.class).processingFinished());
```
Comment (mbaechler) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java]: please always prefer untilAsserted

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 4]
```diff
@@ -89,6 +94,45 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
...
+        Port smtpPort = server.getProbe(SmtpGuiceProbe.class).getSmtpPort();
+        String message = Resources.toString(Resources.getResource("eml/htmlMail.eml"), StandardCharsets.UTF_8);
+
+        try (SMTPMessageSender sender = new SMTPMessageSender(Domain.LOCALHOST.asString())) {
+            Mono.fromRunnable(
+                Throwing.runnable(() -> {
+                    sender.connect(JAMES_SERVER_HOST, smtpPort);
+                    sendUniqueMessageToTwoUsers(sender, message);
+                }))
+                .subscribeOn(Schedulers.elastic())
```
Comment (trantienduchn) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java]: It's it(`.subscribeOn(Schedulers.elastic())`) necessary?

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 5]
```diff
@@ -89,6 +94,45 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
 
     }
 
+    @Test
+    default void mailsShouldBeWellReceivedByBothRecipient(GuiceJamesServer server) throws Exception {
```
Comment (trantienduchn) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java]: 1. can you add tests about one hundred mails to multiple recipients? 
2. Two maybe is not great, many more recipients might trigger something different.  Can you try?

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java (hunk 4 above)
Comment (mbaechler) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java]: The whole Mono is probably useless

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java (hunk 5 above)
Comment (remk) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:147]: You are right. It fails if we send on mail to more than 2 recipients...
And with multiple causes : cassandra concurrent exception when saving the message to all the recipients,
interrupted exception in StoreMessageManager.createAndDispatchMessage
, null pointer during spamassassin

Comment (remk) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:150]: I added a test with 10 recipients. It's slow but it works. The errors i reported previously where due to test timeout.

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 6]
```diff
@@ -189,4 +256,18 @@ default void sendUniqueMessage(SMTPMessageSender sender, String message) throws
...
+        sendUniqueMessageToUsers(sender, message, ImmutableList.of(JAMES_USER, OTHER_USER));
+    }
+
+    default void sendUniqueMessageToUsers(SMTPMessageSender sender, String message, ImmutableList<String> users) throws IOException {
+        String uniqueMessage = message.replace("banana", "UUID " + UUID.randomUUID().toString());
+        sender.sendMessageWithHeaders("bob@apache.org", users, uniqueMessage);
+    }
+
+    default ImmutableList<String> generateNUsers(int nbUsers) {
+        return IntStream.range(0, nbUsers).boxed().map(index -> "user" + index + "@" + DOMAIN).collect(Guavate.toImmutableList());
```
Comment (blackheaven) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:270]: split in lines

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 7]
```diff
@@ -150,6 +146,77 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
...
+        }
+
+        CALMLY_AWAIT_FIVE_MINUTE.untilAsserted(() -> assertThat(server.getProbe(SpoolerProbe.class).processingFinished()).isTrue());
+
+        try (IMAPMessageReader reader = new IMAPMessageReader()) {
+            users.forEach((ThrowingConsumer<String>) user -> reader
+                    .connect(JAMES_SERVER_HOST, server.getProbe(ImapGuiceProbe.class).getImapPort())
+                    .login(user, PASSWORD)
+                    .select(IMAPMessageReader.INBOX)
+                    .awaitMessageCount(CALMLY_AWAIT, 1));
```
Comment (blackheaven) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:216]: indent

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 8]
```diff
@@ -150,6 +146,77 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
...
+
+        Port smtpPort = server.getProbe(SmtpGuiceProbe.class).getSmtpPort();
+        String message = Resources.toString(Resources.getResource("eml/htmlMail.eml"), StandardCharsets.UTF_8);
+
+        try (SMTPMessageSender sender = new SMTPMessageSender(Domain.LOCALHOST.asString())) {
+            sender.connect(JAMES_SERVER_HOST, smtpPort);
+            sendUniqueMessageToUsers(sender, message, users);
+        }
+
+        CALMLY_AWAIT_FIVE_MINUTE.untilAsserted(() -> assertThat(server.getProbe(SpoolerProbe.class).processingFinished()).isTrue());
```
Comment (blackheaven) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:209]: split in lines

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java [hunk 9]
```diff
@@ -150,6 +146,77 @@ default void mailsShouldBeWellReceived(GuiceJamesServer server) throws Exception
...
+
+        ImmutableList<String> users = generateNUsers(10);
+
+        users.forEach((ThrowingConsumer<String>) user -> server.getProbe(DataProbeImpl.class)
+                        .fluent()
+                        .addUser(user, PASSWORD));
+
+        MailboxProbeImpl mailboxProbe = server.getProbe(MailboxProbeImpl.class);
+        mailboxProbe.createMailbox("#private", JAMES_USER, DefaultMailboxes.INBOX);
+        users.forEach((ThrowingConsumer<String>) user ->
```
Comment (blackheaven) [inline server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:197]: why do you have to cast it?

Comment (remk) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:197]: to be able to use a consumer which can throw in the lambda.

Comment (blackheaven) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:197]: is it not inferred?

**Comment by chibenwa** [comment]
```
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project james-server-guice-common: Compilation failure: Compilation failure: 
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[170,14] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[170,45] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[173,25] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[177,25] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[216,14] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[216,45] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[220,25] cannot find symbol
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
[b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] -> [Help 1]
```

File: server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java (hunk 9 above)
Comment (remk) [reply server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:197]: no :(

**Comment by remk** [comment]
> ```
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project james-server-guice-common: Compilation failure: Compilation failure: 
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[170,14] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[170,45] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[173,25] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[177,25] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[216,14] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[216,45] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   class IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] /james-project/server/container/guice/guice-common/src/test/java/org/apache/james/MailsShouldBeWellReceived.java:[220,25] cannot find symbol
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   symbol:   variable IMAPMessageReader
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR]   location: interface org.apache.james.MailsShouldBeWellReceived
> [b73b03ac15f9b24a04ca6b25ba8c9ca8420f9769] [ERROR] -> [Help 1]
> ```

Strange, compile on local, and the import were not changed

**Comment by remk** [comment]
test this please

**Comment by chibenwa** [comment]
try a rebase on master maybe?

**Comment by remk** [comment]
> try a rebase on master maybe?

That was it. ImapReader was renamed to TestIMAPClient

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**